sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala

```diff
@@ -228,4 +222,152 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext {
+    testRead(spark.read.csv(Seq(dir, dir): _*), data ++ data, schema)
+    // Test explicit calls to single arg method - SPARK-16009
+    testRead(Option(dir).map(spark.read.csv).get, data, schema)
+
+    // Reader, with user specified schema, should just apply user schema on the file data
+    testRead(spark.read.schema(userSchema).csv(), Seq.empty, userSchema)
+    testRead(spark.read.schema(userSchema).csv(dir), data, userSchema)
+    testRead(spark.read.schema(userSchema).csv(dir, dir), data ++ data, userSchema)
+    testRead(spark.read.schema(userSchema).csv(Seq(dir, dir): _*), data ++ data, userSchema)
+  }
+
+  test("json - API and behavior regarding schema") {
+    // Writer
+    spark.createDataset(data).toDF("str").write.mode(SaveMode.Overwrite).json(dir)
+    val df = spark.read.json(dir)
+    checkAnswer(df, spark.createDataset(data).toDF())
+    val schema = df.schema
+
+    // Reader, without user specified schema
+    intercept[AnalysisException] {
+      testRead(spark.read.json(), Seq.empty, schema)
+    }
+    testRead(spark.read.json(dir), data, schema)
+    testRead(spark.read.json(dir, dir), data ++ data, schema)
+    testRead(spark.read.json(Seq(dir, dir): _*), data ++ data, schema)
+    // Test explicit calls to single arg method - SPARK-16009
+    testRead(Option(dir).map(spark.read.json).get, data, schema)
+
+    // Reader, with user specified schema, data should be nulls as schema in file different
+    // from user schema
+    val expData = Seq[String](null, null, null)
+    testRead(spark.read.schema(userSchema).json(), Seq.empty, userSchema)
+    testRead(spark.read.schema(userSchema).json(dir), expData, userSchema)
+    testRead(spark.read.schema(userSchema).json(dir, dir), expData ++ expData, userSchema)
+    testRead(spark.read.schema(userSchema).json(Seq(dir, dir): _*), expData ++ expData, userSchema)
+  }
+
+  test("parquet - API and behavior regarding schema") {
+    // Writer
+    spark.createDataset(data).toDF("str").write.mode(SaveMode.Overwrite).parquet(dir)
+    val df = spark.read.parquet(dir)
+    checkAnswer(df, spark.createDataset(data).toDF())
+    val schema = df.schema
+
+    // Reader, without user specified schema
+    intercept[AnalysisException] {
+      testRead(spark.read.parquet(), Seq.empty, schema)
+    }
+    testRead(spark.read.parquet(dir), data, schema)
+    testRead(spark.read.parquet(dir, dir), data ++ data, schema)
+    testRead(spark.read.parquet(Seq(dir, dir): _*), data ++ data, schema)
+    // Test explicit calls to single arg method - SPARK-16009
+    testRead(Option(dir).map(spark.read.parquet).get, data, schema)
+
+    // Reader, with user specified schema, data should be nulls as schema in file different
+    // from user schema
+    val expData = Seq[String](null, null, null)
+    testRead(spark.read.schema(userSchema).parquet(), Seq.empty, userSchema)
+    testRead(spark.read.schema(userSchema).parquet(dir), expData, userSchema)
```

Comment:
Let's documents that as a test for now.
On Jun 21, 2016 7:51 PM, "Xiao Li" <notifications@github.com> wrote:

> In
> sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala
> <https://github.com/apache/spark/pull/13727#discussion_r67987384>:
>
> > +
> > +    // Reader, without user specified schema
> > +    intercept[AnalysisException] {
> > +      testRead(spark.read.parquet(), Seq.empty, schema)
> > +    }
> > +    testRead(spark.read.parquet(dir), data, schema)
> > +    testRead(spark.read.parquet(dir, dir), data ++ data, schema)
> > +    testRead(spark.read.parquet(Seq(dir, dir): _*), data ++ data, schema)
> > +    // Test explicit calls to single arg method - SPARK-16009
> > +    testRead(Option(dir).map(spark.read.parquet).get, data, schema)
> > +
> > +    // Reader, with user specified schema, data should be nulls as schema in file different
> > +    // from user schema
> > +    val expData = Seq[String](null, null, null)
> > +    testRead(spark.read.schema(userSchema).parquet(), Seq.empty, userSchema)
> > +    testRead(spark.read.schema(userSchema).parquet(dir), expData, userSchema)
>
> @tdas <https://github.com/tdas> ORC behaves differently. When the
> user-specified schema does not match the physical schema, it simply stops
> and reports an exception. Do you think that behavior is better than
> returning null for all the rows?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/spark/pull/13727/files/2539a947d382d8d6c21c59fe8a9420a15aad9b9a#r67987384>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAoerKNkJhd5r8QzCyrTjIY8aMnl9uVZks5qOKMygaJpZM4I4AVc>
> .
>